Datadog integration

AmineChikhaoui opened the pull request:
The purpose of this PR is to integrate a Datadog backend into Nixops. The end goal is to be able to deploy monitors/timeboards as resources in the same fashion as tools like [Terraform](https://www.terraform.io/docs/providers/datadog/index.html).

It's still a work in progress. Next steps needed are:

- [x] add all the supported options by the datadog api to the monitor options.
- [x] Implement a timeboard resource
- [x] Implement a screenboard resource
- [x] add examples
- [x] update the nixops manual if needed, etc ..

cc @rbvermaa

rbvermaa commented on nix/eval-machine-info.nix:
```diff
@@ -99,6 +99,10 @@ rec {
   resources.machines = mapAttrs (n: v: v.config) nodes;
 
+  # Datadog resources
+
```
Please remove extra empty line.

rbvermaa commented on nixops/resources/datadog-monitor.py:
```diff
@@ -0,0 +1,193 @@
+    """State of a Datadog monitor"""
+
+    state = nixops.util.attr_property("state", nixops.resources.ResourceState.MISSING, int)
+    apiKey = nixops.util.attr_property("apiKey", None)
```
In python code, we usually use _'s and not camelCase for properties of python classes.

rbvermaa commented on nix/datadog-monitor.nix:
```diff
@@ -0,0 +1,111 @@
+          };
+        };
+  config = {
+    renotify_interval = mkIf (builtins.isString config.escalationMessage) (throw "renotifyInterval can't be used when the escalationMessage is set.");
```
I think it might be better to use the 'assertions'. For examples, see nixpkgs/nixos, e.g. in the xserver.nix NixOS module.

AmineChikhaoui replied:
@rbvermaa I finally removed this check as it's not really needed. Turned out the api handles all cases without issues from a couple of tests I did.

AmineChikhaoui commented:
@rbvermaa and @edolstra , can you please review this. The changes that I added are: removing all xml parsing and I modified the datadog monitor options to be one string json option (see example) instead of defining them one by one which will require maintaining it frequently whenever the datadog api is changed.

rbvermaa commented:
@AmineChikhaoui did you update the example?

AmineChikhaoui commented:
@rbvermaa yes see examples/datadog-monitor.nix

rbvermaa commented:
@AmineChikhaoui Ah thnx, I missed the builtins.toJSON :)

rbvermaa commented:
@AmineChikhaoui could we default to using an environment variable for the app and api key of datadog?

AmineChikhaoui commented:
@rbvermaa Yes I'll push a change for that

AmineChikhaoui commented on nixops/datadog_utils.py:
```diff
@@ -0,0 +1,24 @@
+    if not api_key: api_key = os.environ.get('DATADOG_API_KEY')
+    if not app_key: app_key = os.environ.get('DATADOG_APP_KEY')
+    if not api_key or not app_key:
+        raise Exception("please set the datadog apiKey and appKey options (or the environment variables DATADOG_API_KEY and DATADOG_APP_KEY)")
```
@rbvermaa done, see the changes above.

AmineChikhaoui commented:
@rbvermaa I added a minimal guide on how to deploy the resources, let me know if you have any comments. thanks
